src/python/xml_parser-py.c

```diff
@@ -792,3 +790,247 @@ py_xml_parse_main_metadata_together(G_GNUC_UNUSED PyObject *self, PyObject *args
 
     Py_RETURN_NONE;
 }
+
+typedef struct {
+    PyObject_HEAD
+    cr_PkgIterator *pkg_iterator;
+} _PkgIteratorObject;
+
+cr_PkgIterator *
+PkgIterator_FromPyObject(PyObject *o)
+{
+    if (!PkgIteratorObject_Check(o)) {
+        PyErr_SetString(PyExc_TypeError, "Expected a createrepo_c.PkgIterator object.");
+        return NULL;
+    }
+    return ((_PkgIteratorObject *)o)->pkg_iterator;
+}
+
+PyObject *
+Object_FromPkgIterator(cr_PkgIterator *pkg_iterator)
+{
+    if (!pkg_iterator) {
+        PyErr_SetString(PyExc_ValueError, "Expected a cr_PkgIterator pointer not NULL.");
+        return NULL;
+    }
+
+    PyObject *py_pkg_iterator = PyObject_CallObject((PyObject *)&PkgIterator_Type, NULL);
+    ((_PkgIteratorObject *)py_pkg_iterator)->pkg_iterator = pkg_iterator;
+    return py_pkg_iterator;
+}
+
+static int
+check_PkgIteratorStatus(const _PkgIteratorObject *self)
+{
+    assert(self != NULL);
+    assert(PkgIteratorObject_Check(self));
+    if (self->pkg_iterator == NULL) {
+        PyErr_SetString(CrErr_Exception, "Improper createrepo_c PkgIterator object.");
+        return -1;
+    }
+    return 0;
+}
+
+/* Function on the type */
+
+static PyObject *
+pkg_iterator_new(PyTypeObject *type,
+            G_GNUC_UNUSED PyObject *args,
+            G_GNUC_UNUSED PyObject *kwds)
+{
+    _PkgIteratorObject *self = (_PkgIteratorObject *)type->tp_alloc(type, 0);
+    if (self) {
+        self->pkg_iterator = NULL;
+    }
+    return (PyObject *)self;
+}
+
+PyDoc_STRVAR(pkg_iterator_init__doc__,
+             "PkgIterator object\n\n"
+             ".. method:: __init__()\n\n"
+             "    Default constructor\n");
+
+static int
+pkg_iterator_init(_PkgIteratorObject *self, PyObject *args, PyObject *kwargs)
+{
+    char *primary_path;
+    char *filelists_path;
+    char *other_path;
+    PyObject *py_newpkgcb, *py_warningcb;
+    CbData cbdata;
```

Comment:
This is likely the biggest source of the segfaults.
This `cbdata` is valid only in the scope of `pkg_iterator_init` but we save a pointer to it in `cr_PkgIterator_new()`.
Since we need it through out the parsing I think we have allocate it on the heap and store a pointer in `_PkgIteratorObject`.
```suggestion
```